### PR TITLE
Login: Add verify with code handling for 2FA

### PR DESCRIPTION
**Pull request by gziolo**
This is how it looks on production at the moment:

screen_shot_2017-05-09_at_14 15 47

Calypso version which needs some more love:

screen shot 2017-05-09 at 15 05 20

### Testing

1. Go to http://calypso.localhost:3000/login.
2. Use credentials for your account that has 2FA based on the Authenticator app enabled and WordPress.com app installed on the same device.
3. Once you get push notification click `Verify with code instead` button.
4. Enter code from the Authenticator app.
5. You should login.

### Review

* [x] Code
* [x] Product

### CODE REVIEW

File: client/state/login/reducer.js [hunk 1]
```diff
@@ -48,6 +49,7 @@ export const twoFactorAuth = createReducer( null, {
 	[ TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_FAILURE ]: updateTwoStepNonce,
 	[ TWO_FACTOR_AUTHENTICATION_SEND_SMS_CODE_REQUEST_FAILURE ]: updateTwoStepNonce,
 	[ TWO_FACTOR_AUTHENTICATION_SEND_SMS_CODE_REQUEST_SUCCESS ]: updateTwoStepNonce,
+	[ TWO_FACTOR_AUTHENTICATION_LOGIN_VERIFY_WITH_CODE ]: ( state ) => ( { ...state, two_step_notification_sent: null } ),
```
Comment (gziolo) [inline client/state/login/reducer.js]: Need to double check what to set here.

Comment (gziolo) [inline client/state/login/reducer.js]: `none` it is :)

**Comment by drewblaisdell** [comment]
The code looks good to me.

There is an issue with login right now that is relevant to the code in this PR: we don't associate the different screens in the login process with routes, so:

- The back button does not behave as expected, e.g. clicking the native browser 'Back' from any 2FA screen does not return the user to the login page.
- We need to make unintuitive changes to state when we want to change screens. In the case of this PR, changing `two_step_notification_sent` to `none` in order to change screens works, but what if the user wants to authenticate using the app after they click the `Verify with code instead` button?

I think ideally:

- The top-level login component should be mounted on distinct routes, e.g. `/login`, `/login/code`, `/login/app`.
- Once a push notification is sent, both `/login/code` and `/login/app` should poll the server to see if the user approved the login on their phone.

Neither point is something we should address in this PR, and I think this solution is good for now, but I'm curious what you think.

**Comment by fabianapsimoes** [comment]
Product :+1: 

Question: when will the "recovery text" link be addressed? Do we have a separate task for that?

**Comment by gziolo** [comment]
We agreed to tackle or remaining issues in the follow up PR. We had similar concerns as @drewblaisdell mentioned when we discussed the current state of art yesterday. I'm going to document all different 2FA configurations and then we can pick the best way to solve navigation and corresponding links. Thanks for great feedback everyone!

**Pull request merged**